http.go

```diff
@@ -0,0 +1,455 @@
+		return setLibgit2Error(err)
+	}
+
+	managed, err := transport.Action(C.GoString(url), SmartService(action))
+	if err != nil {
+		return setLibgit2Error(err)
+	}
+
+	stream := C.calloc(1, C.size_t(unsafe.Sizeof(C.managed_smart_subtransport_stream{})))
+	managedPtr := pointerHandles.Track(managed)
+	C._go_git_setup_smart_subtransport_stream((*C.managed_smart_subtransport_stream)(stream), managedPtr)
+
+	*out = (*C.git_smart_subtransport_stream)(stream)
+	return 0
+}
+
+//export httpClose
+func httpClose(t *C.git_smart_subtransport) C.int {
+	transport, err := getSmartSubtransportInterface(t)
+	if err != nil {
+		return setLibgit2Error(err)
+	}
+
+	if err := transport.Close(); err != nil {
+		return setLibgit2Error(err)
+	}
+
+	return 0
+}
+
+//export httpFree
+func httpFree(transport *C.git_smart_subtransport) {
+	wrapperPtr := (*C.managed_smart_subtransport)(unsafe.Pointer(transport))
+	pointerHandles.Untrack(wrapperPtr.ptr)
+}
+
+var errNoSmartSubtransport = errors.New("passed object does not implement SmartSubtransport")
+
+func getSmartSubtransportInterface(_t *C.git_smart_subtransport) (SmartSubtransport, error) {
+	wrapperPtr := (*C.managed_smart_subtransport)(unsafe.Pointer(_t))
+
+	transport, ok := pointerHandles.Get(wrapperPtr.ptr).(SmartSubtransport)
+	if !ok {
+		return nil, errNoSmartSubtransport
+	}
+
+	return transport, nil
+}
+
+//export httpTransportCb
+func httpTransportCb(out **C.git_transport, owner *C.git_remote, param unsafe.Pointer) C.int {
+	return C._go_git_transport_smart(out, owner)
+}
+
+//export httpSmartSubtransportCb
+func httpSmartSubtransportCb(out **C.git_smart_subtransport, owner *C.git_transport, param unsafe.Pointer) C.int {
+	if out == nil {
+		return -1
+	}
+
+	transport := C.calloc(1, C.size_t(unsafe.Sizeof(C.managed_smart_subtransport{})))
+	managed := &ManagedTransport{owner: owner}
+	managedPtr := pointerHandles.Track(managed)
+	C._go_git_setup_smart_subtransport((*C.managed_smart_subtransport)(transport), managedPtr)
+
+	*out = (*C.git_smart_subtransport)(transport)
+	return 0
+}
+
+var errNoSmartSubtransportStream = errors.New("passed object does not implement SmartSubtransportStream")
+
+func getSmartSubtransportStreamInterface(_s *C.git_smart_subtransport_stream) (SmartSubtransportStream, error) {
+	wrapperPtr := (*C.managed_smart_subtransport_stream)(unsafe.Pointer(_s))
+
+	transport, ok := pointerHandles.Get(wrapperPtr.ptr).(SmartSubtransportStream)
+	if !ok {
+		return nil, errNoSmartSubtransportStream
+	}
+
+	return transport, nil
+}
+
+//export smartSubtransportRead
+func smartSubtransportRead(s *C.git_smart_subtransport_stream, data *C.char, l C.size_t, read *C.size_t) C.int {
+	stream, err := getSmartSubtransportStreamInterface(s)
+	if err != nil {
+		return setLibgit2Error(err)
+	}
+
+	var p []byte
+	header := (*reflect.SliceHeader)(unsafe.Pointer(&p))
+	header.Cap = int(l)
+	header.Len = int(l)
+	header.Data = uintptr(unsafe.Pointer(data))
+
+	n, err := stream.Read(p)
+	if err != nil {
```

Comment:
Per https://golang.org/pkg/io/#Reader, this should be

```suggestion
	if n == 0 && err != nil {
```